exporter/clickhouseexporter/internal/metrics_model.go

```
@@ -0,0 +1,467 @@
+	}
+}
+
+func InsertMetrics(ctx context.Context, tx *sql.Tx, metricsMap map[pmetric.MetricType]MetricsModel, logger *zap.Logger) error {
+	var errs []error
+	for _, metrics := range metricsMap {
+		errs = append(errs, metrics.Insert(ctx, tx, logger))
+	}
+	return multierr.Combine(errs...)
+}
+
+func convertExemplars(exemplars pmetric.ExemplarSlice) (clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet) {
+	var (
+		attrs       clickhouse.ArraySet
+		times       clickhouse.ArraySet
+		floatValues clickhouse.ArraySet
+		intValues   clickhouse.ArraySet
+		traceIDs    clickhouse.ArraySet
+		spanIDs     clickhouse.ArraySet
+	)
+	for i := 0; i < exemplars.Len(); i++ {
+		exemplar := exemplars.At(i)
+		if !exemplar.TraceID().IsEmpty() && !exemplar.SpanID().IsEmpty() {
+			attrs = append(attrs, attributesToMap(exemplar.FilteredAttributes()))
+			times = append(times, exemplar.Timestamp().AsTime())
+			floatValues = append(floatValues, exemplar.DoubleValue())
+			intValues = append(intValues, exemplar.IntValue())
+			traceID := exemplar.TraceID()
+			traceIDs = append(traceIDs, hex.EncodeToString(traceID[:]))
+			spanID := exemplar.SpanID()
+			spanIDs = append(spanIDs, hex.EncodeToString(spanID[:]))
+		}
+	}
+	return attrs, times, floatValues, intValues, traceIDs, spanIDs
+}
+
+func attributesToMap(attributes pcommon.Map) map[string]string {
+	m := make(map[string]string, attributes.Len())
+	attributes.Range(func(k string, v pcommon.Value) bool {
+		m[k] = v.Str()
+		v.Type()
+		return true
+	})
+	return m
+}
+
+func convertSliceToArraySet(slice interface{}) clickhouse.ArraySet {
+	var set clickhouse.ArraySet
+	switch slice := slice.(type) {
+	case []uint64:
+		for _, item := range slice {
+			set = append(set, item)
+		}
+	case []float64:
+		for _, item := range slice {
+			set = append(set, item)
+		}
+	default:
+		fmt.Printf("unsupported slice type: %s", reflect.TypeOf(slice))
```

you should return an error instead. You cannot Printf in code. If you must output, pass a logger in as argument.

---

done